custom/icds/messaging/indicators.py

```diff
@@ -382,27 +387,34 @@ def get_messages(self, language_code=None):
 
 
 def get_awcs_with_old_vhnd_date(domain, awc_location_ids):
-    return set(awc_location_ids) - precompute_awws_in_vhnd_timeframe(domain)
+    return set(awc_location_ids) - get_awws_in_vhnd_timeframe(domain)
 
 
 @icds_quickcache(timeout=12 * 60 * 60, memoize_timeout=12 * 60 * 60, session_function=None)
-def precompute_awws_in_vhnd_timeframe(domain):
+def get_awws_in_vhnd_timeframe(domain):
+    # This function is called concurrently by many tasks.
+    # The CriticalSection ensures that the expensive operation is not triggered
+    #   by each task again, the compute_awws_in_vhnd_timeframe itself is cached separately,
+    #   so that other waiting tasks could lookup the value from cache.
+    with CriticalSection(['compute_awws_in_vhnd_timeframe']):
+        return compute_awws_in_vhnd_timeframe(domain)
+
+
+@icds_quickcache(timeout=60 * 60, memoize_timeout=60 * 60, session_function=None)
+def compute_awws_in_vhnd_timeframe(domain):
     """
     This computes awws with vhsnd_date_past_month less than 37 days.
 
     Result is cached in local memory, so that indvidual reminder tasks
     per AWW/LS dont hit the database each time
     """
-    from corehq.apps.userreports.util import get_table_name
     table = get_table_name(domain, 'static-vhnd_form')
     query = """
     SELECT DISTINCT awc_id
     FROM "{table}"
-    WHERE '{today}' - max(vhsnd_date_past_month) < 37
-    """.format(table=table, today=datetime.today().date())
+    WHERE vhsnd_date_past_month > '{{old}}'
+    """.format(table=table, old=datetime.today().date() - timedelta(days=37))
```

Comment:
I think that this may work, but it isn't recommended. For tables, you do need to pass the variable into a formatted string, but for parameters its generally recommended to pass this as a param to the execute function:

```python
query = "... WHERE vhsnd_date_past_month > %(37_days_ago)s"
query_params = {"37_days_ago": datetime.today().date() - timedelta(days=37))

with connection as cursor:
    cursor.execute(query, query_params)
```

The reason to do this is because `execute` knows how to escape parameters and properly format them for queries. That's super important for user input, less important for input that we control, but still a good practice